### PR TITLE
For #54: created card effects and actions

### LINKED ISSUE #54: Backend: Begin Designing Composable Card Effects
Card should be able to add custom effects.

Initially, should support basic effects like:
- Deal X damage to Single Opponent.
- Deal X damage to Every Opponent.
- Apply Buff for X Turns to target.
- Heal X HP to Single Ally.

Cards should be able to be created as Assets in the Editor. Should be able to add/remove effects.

### PR DESCRIPTION
**Pull request by paulodamaso**
Closes #54 

- created actions for heal, buff and damage
- created effects for HealSingle, HealAll, DamageSingle and BuffSingle
- due to lack of a Target element (the object to which we will apply the effect), the actual effects scripts were not implemented. Left a puzzle to create the effects the target element and then the effects implementations

### CODE REVIEW

File: src/ProjectNeon/Assets/Cards/Effect.cs [hunk 1]
```diff
@@ -3,4 +3,6 @@
 [CreateAssetMenu()]
 public class EffectAction : ScriptableObject
 {
+    [SerializeField]
+    private ScriptableObject effect;
```
Comment (SilasReinagel) [inline src/ProjectNeon/Assets/Cards/Effect.cs]: How will this actually work in the code?

If it's just a scriptable object, then we will have to use Reflection or something in order to actually execute it.

This design will probably need to be more like Command Pattern, where each effect is a Command.

File: src/ProjectNeon/Assets/Cards/Effects/Actions/BuffAction.cs [hunk 2]
```diff
@@ -0,0 +1,10 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+[CreateAssetMenu()]
+public class BuffAction : ScriptableObject
```
Comment (SilasReinagel) [inline src/ProjectNeon/Assets/Cards/Effects/Actions/BuffAction.cs]: How will we know what kind of buff to apply?

Will this increase Physical Damage amounts?
Will this increase Armor?
Will this let a character play an extra card?

We need some design that will support Polymorphism.

**Comment by SilasReinagel** [comment]
@paulodamaso As design insight, we probably just need a Source Character and a List of Target Characters in order to execute an action. 

Effect + Source + Targets, then we can call "Execute".

This will be especially relevant, since maybe Character A have better Ranged Weapon skills than Character B, so it might use Character A Ranged Weapon Stat to calculate damage.

**Comment by 0crat** [comment]
@silasreinagel[/z](https://www.0crat.com/u/silasreinagel) please, pay attention to this pull request

**Comment by 0crat** [comment]
Job [#122](https://github.com/enigmadragons/projectneon/issues/122) is now in scope, role is `REV`

File: src/ProjectNeon/Assets/Cards/Effect.cs (hunk 1 above)
Comment (paulodamaso) [reply src/ProjectNeon/Assets/Cards/Effect.cs]: @SilasReinagel Actually I didn't understood quite well `EffectAction` class significance into the project. The concept is clear for me, just its implementation as a class makes me puzzled. The `effect` variable exists here just due to the fact that I wasn't able to extend / implement `EffectAction` class in a polymorphic way using UnityEditor; I wasn't even able to add scripts to it using just the UI, so I've created this field so I could put scripts into `EffectAction`. I don't know how some concepts are transposed to Unity universe, but in 'traditional OO' I'd create an Interface `EffectAction` and implement it as many times as desired for each effect we have. Wouldn't be better if we had `EffectAction` as an interface and follow this proposed logic?

File: src/ProjectNeon/Assets/Cards/Effects/Actions/BuffAction.cs (hunk 2 above)
Comment (paulodamaso) [reply src/ProjectNeon/Assets/Cards/Effects/Actions/BuffAction.cs]: @SilasReinagel I think that Buff cards are too complex to model this early; we barely have character attributes and enemy stats defined, so creating these effects so early in project would just make we have to refactor them later with no significant gains for making it now. I propose to evolve basic card and effects behavior and leave a puzzle for ceating buffs in the future

**Comment by paulodamaso** [comment]
@SilasReinagel Talking about design, wouldn't be better if we don't discriminate between character / enemy in card effects? For example, `Heal Single` card should have the same behavior doesn't matter if player or enemy use it: it should heal some target, regardless if its an enemy or player character. This would allow us to make more generic cards and allow them to be reused. We should continue to check if target is ally / oponent though.

**Comment by 0crat** [comment]
@silasreinagel[/z](https://www.0crat.com/u/silasreinagel) everybody who has role `REV` is banned at [#122](https://github.com/enigmadragons/projectneon/issues/122); I won't be able to assign anyone automatically; consider assigning someone manually (as in [§19](http://www.zerocracy.com/policy.html#19)), or invite more people (as in [§51](http://www.zerocracy.com/policy.html#51)), or remove the job from the scope (as in [§14](http://www.zerocracy.com/policy.html#14))

**Comment by SilasReinagel** [comment]
@paulodamaso Yes, absolutely. We don't need to discriminate.

**Comment by 0crat** [comment]
@silasreinagel[/z](https://www.0crat.com/u/silasreinagel) everybody who has role `REV` is banned at [#122](https://github.com/enigmadragons/projectneon/issues/122); I won't be able to assign anyone automatically; consider assigning someone manually (as in [§19](http://www.zerocracy.com/policy.html#19)), or invite more people (as in [§51](http://www.zerocracy.com/policy.html#51)), or remove the job from the scope (as in [§14](http://www.zerocracy.com/policy.html#14))

**Comment by 0crat** [comment]
The job [#122](https://github.com/enigmadragons/projectneon/issues/122) is now out of scope

**Comment by 0crat** [comment]
Pull request [#122](https://github.com/enigmadragons/projectneon/issues/122) was not merged, no payment for `ARC`, see [§28](http://www.zerocracy.com/policy.html#28)

**Comment by paulodamaso** [comment]
@SilasReinagel Closed thios one because we already have #177

**Pull request closed without merging**